### ISSUE TITLE
Tracking: Russian users unable to connect to Modrinth's meta servers with the app

### ISSUE THREAD

**Issue by brawaru** [modrinth/theseus#858]
### Describe the bug

When Modrinth launches, it has to connect to `launcher-files.modrinth.com` in order to detect internet connection, and then `meta.modrinth.com` to fetch manifests containing Minecraft and mod loaders versions. When any of this fails, Modrinth app will refuse to launch with ‘Unable to read launcher metadata from any source’ error.

Recently, we had multiple tickets opened by Russian users where their app would refuse to launch exactly with this error and logs would either contain an error related to fetching of [minecraft/manifest.json](https://meta.modrinth.com/minecraft/v0/manifest.json) file, or Modrinth failing to read cache because previous request to detect internet connection has silently failed, and app is operating in offline mode.

Otherwise, the app is working as usual when using a VPN.

### Steps to reproduce

1. Remove the Modrinth app cache (`%AppData%\com.modrinth.theseus\caches`).
2. Start the Modrinth app.

### Expected behavior

Modrinth app is able to connect to its meta server and launch normally.

### System information

Most reports are from Windows.

### Additional context

My initial assumption is that this error is related to Russian internet censorship. Modrinth relies on Cloudflare to protect its network services, and various sites deemed illegal in Russia are also using have used Cloudflare. When Russian network watchdog ([Roskomnadzor](https://en.wikipedia.org/wiki/Roskomnadzor)) bans internet resources, they ban the URL deemed illegal if it's an HTTP resource, or the whole domain name and IPs associated with it if that's an HTTPS resource.

Cloudflare works by sharing the IP addresses between the domains, and then proxying the connection to the right service based on the domain name in a request, so domains sharing IPs that were banned by Roskomnadzor may end up in a collateral damage. Other parties have attempted contacting Roskomnadzor to unblock those IPs, but these attempts have been met with hostility and dismissiveness towards the role of Cloudflare as a protection service, where Roskomnadzor recommended not using it, and in some cases even suggested removing HTTPS protection to allow them to selectively block illegal resources. Which is obviously not a solution and no one ever should do that.

However, to minimise the damage, a Deep Package Inspection (DPI) solution is applied that checks the unencrypted HTTPS handshakes and lets connections to other domain names through, despite the IP ban. If DPI fails to determine the domain name, the connection is dropped right during the handshake, which shows as `PR_CONNECT_RESET_ERROR` in browsers.

Weirdly enough, users reporting this issue seem to be fully able to fetch the remote meta files in their browsers or using PowerShell script, which may mean that Modrinth app is doing something in an extraordinary way when it comes to handshake, which seems to trip DPI.

Unfortunately, Modrinth app is missing any logging related to the actual network error, so it's all just a theory and we cannot tell for sure what's going on. For all we know, it can also be just Cloudflare itself challenging the Modrinth app network client. This is why this issue is marked as ‘tracking’: we need more information.

Modrinth needs to implement logging for this and other network errors, which would help to diagnose this issue. Russian users experiencing this error and having knowledge of tools like HTTP Toolkit and Wireshark are welcome to diagnose this issue on their own and share their findings with us here.

Please make sure that this issue is not relevant to cache corruption, which is a separate issue. To check this:
- **Modrinth app fails to detect internet connection**
  - `caches` directory is missing from `%AppData%\com.modrinth.theseus`
  - The log contains the following:
    ```
    INFO theseus_gui: Initialized tracing subscriber. Loading Modrinth App!
    INFO init_watcher: theseus::state: Initting watcher
    ERROR theseus_gui::error: error=Unable to read launcher tags from any source span_trace=   0: theseus::state::tags::init
           with dirs=... fetch_online=false credentials=...
    ```
- **Modrinth app fails to fetch metadata**
  - The log contains the following:
    ```
    INFO theseus_gui: Initialized tracing subscriber. Loading Modrinth App!
    INFO init_watcher: theseus::state: Initting watcher
    WARN initialize_state:initialize_state:init{dirs=... fetch_online=true}: theseus::state::metadata: Unable to fetch launcher metadata: Metadata error: Unable to fetch https://meta.modrinth.com/minecraft/v0/manifest.json
    ERROR theseus_gui::error: error=Unable to read launcher metadata from any source span_trace=   0: theseus::state::metadata::init
           with dirs=... fetch_online=true
             at theseus\src\state\metadata.rs:63
    1: theseus::state::initialize_state
             at theseus\src\state\mod.rs:134
    2: theseus_gui::initialize_state
             at theseus_gui\src-tauri\src\main.rs:16
    ```

**Comment by brawaru** [comment]
This seems to be resolved by the previously affected ISPs. We have also put up a page on resolving issues like this one at https://support.modrinth.com/en/articles/8849048.

**Comment by VadimP22** [comment]
In my case, this is a problem with the Windows version of the application. Everything works fine in Linux. So I don’t think it’s a matter of the censorship system.
